### PR TITLE
Use native_gate_virtualz to handle iSWAP or CZ

### PR DESCRIPTION
**Pull request by Jacfomg**
It makes the virtualz calibration routine able to take both iSWAP and CZ. It comes from #568. I just separed them to ease the review. It comes from an old routine so I will update it if you think this is interesting.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.
- [ ] Compatibility with Qibo modules (Please edit this section if the current pull request is not compatible with the following branches).
    - [ ] Qibo: `master`
    - [ ] Qibolab: `main`
    - [ ] Qibolab_platforms_qrc: `main`

### CODE REVIEW

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 1]
```diff
@@ -32,47 +32,45 @@ class CZVirtualZParameters(Parameters):
...
-    """CzVirtualZ outputs when fitting will be done."""
+class VirtualZResults(Results):
+    """VirtualZ outputs when fitting will be done."""
 
     fitted_parameters: dict[tuple[str, QubitId],]
     """Fitted parameters"""
-    cz_angle: dict[QubitPairId, float]
-    """CZ angle."""
+    native_gate_angle: dict[QubitPairId, float]
+    """Angle."""
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:51]:
```suggestion
    """Native gate angle."""
```

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 2]
```diff
@@ -32,47 +32,45 @@ class CZVirtualZParameters(Parameters):
...
 
-    def __contains__(self, key: QubitPairId):
-        """Check if key is in class.
-        While key is a QubitPairId both chsh and chsh_mitigated contain
-        an additional key which represents the basis chosen.
-        """
 
-        return key in [
-            (target, control) for target, control, _ in self.fitted_parameters
-        ]
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:61]: This is necessary, why did you remove it?

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 3]
```diff
@@ -151,17 +158,17 @@ def create_sequence(
...
     )
 
 
 def _acquisition(
-    params: CZVirtualZParameters,
+    params: VirtualZParameters,
     platform: Platform,
-    targets: list[QubitPairId],
-) -> CZVirtualZData:
+    targets: Qubits,
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:168]: Revert this line to the previous version of the code.

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 4]
```diff
@@ -171,14 +178,15 @@ def _acquisition(
     is undone in the high frequency qubit and a theta90 pulse is applied to the low
     frequency qubit before measurement. That is, a pi-half pulse around the relative phase
     parametereized by the angle theta.
-    Measurements on the low frequency qubit yield the 2Q-phase of the gate and the
+    Measurements on the low frequency qubit yield the the 2Q-phase of the gate and the
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:181]: ?

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 5]
```diff
@@ -230,12 +239,11 @@ def _acquisition(
...
-                    (target_q, control_q, setup),
-                    dict(
-                        target=result_target,
-                        control=result_control,
-                    ),
+                    target=target_q,
+                    control=control_q,
+                    setup=setup,
+                    prob_target=result_target,
+                    prob_control=result_control,
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:246]: ?

Comment (andrea-pasquale) [reply src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:246]: The previous version of the code was fine. You can drop even drop the register qubit method.

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 6]
```diff
@@ -260,15 +268,15 @@ def _fit(
...
+    native_gate_angle = {}
     for pair in pairs:
         virtual_phase[pair] = {}
         for target, control, setup in data[pair]:
             target_data = data[pair][target, control, setup].target
             pguess = [
                 np.max(target_data) - np.min(target_data),
                 np.mean(target_data),
-                np.pi,
+                3.14,
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:279]: Why?

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 7]
```diff
@@ -281,62 +289,61 @@ def _fit(
...
-                    target_q, control_q, "I"
-                ][2]
-        except KeyError:
-            log.warning(f"CZ fit failed for pair ({target_q, control_q}).")
-
-    return CZVirtualZResults(
-        cz_angle=cz_angle,
+            except:
+                log.warning("landscape_fit: the fitting was not succesful")
+                fitted_parameters[target, control, setup] = [0] * 3
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:294]:
```suggestion
```

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 8]
```diff
@@ -281,62 +289,61 @@ def _fit(
...
-                virtual_phase[pair][target_q] = -fitted_parameters[
-                    target_q, control_q, "I"
-                ][2]
-        except KeyError:
-            log.warning(f"CZ fit failed for pair ({target_q, control_q}).")
-
-    return CZVirtualZResults(
-        cz_angle=cz_angle,
+            except:
+                log.warning("landscape_fit: the fitting was not succesful")
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:293]: The previous exception was more descriptive.

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 9]
```diff
@@ -281,62 +289,61 @@ def _fit(
...
+            pair,
+            list(pair)[::-1],
+        ):
+            native_gate_angle[target_q, control_q] = abs(
+                fitted_parameters[target_q, control_q, "X"][2]
+                - fitted_parameters[target_q, control_q, "I"][2]
+            )
+            virtual_phase[pair][target_q] = -fitted_parameters[
+                target_q, control_q, "I"
+            ][2]
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:306]: Again, stick to the previous layout including the try-except block.

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 10]
```diff
@@ -281,62 +289,61 @@ def _fit(
...
         cols=2,
         subplot_titles=(
-            f"Qubit {qubits[0]}",
-            f"Qubit {qubits[1]}",
+            f"Qubit {targets[0]}",
+            f"Qubit {targets[1]}",
         ),
     )
-    fitting_report = set()
+    reports = []
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:328]: The code will not work as expected with `[]`, please keep `set()`.

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 11]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
-                    )
+            fitting_report = table_html(
+                table_dict(
+                    [target, target, targets[1]],
+                    [" angle", "Virtual Z phase", "Flux pulse amplitude"],
+                    [
+                        np.round(fit.native_gate_angle[target, control], 4),
+                        np.round(fit.virtual_phase[tuple(sorted(target))][target], 4),
+                        np.round(data.amplitudes[targets]),
+                    ],
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:390]: Go back to previous version after changing list in set

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 12]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
 
     fig1.update_layout(
-        title_text=f"Phase correction Qubit {qubits[0]}",
+        title_text=f"Phase correction Qubit {targets[0]}",
         showlegend=True,
+        uirevision="0",  # ``uirevision`` allows zooming while live plotting
         xaxis1_title="theta [rad] + virtual phase[rad]",
         xaxis2_title="theta [rad] + virtual phase [rad]",
-        yaxis_title="Signal [a.u.]",
+        yaxis_title="MSR[V]",
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:400]: Restore signal

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 13]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
+                        np.round(data.amplitudes[targets]),
+                    ],
                 )
             )
 
     fig1.update_layout(
-        title_text=f"Phase correction Qubit {qubits[0]}",
+        title_text=f"Phase correction Qubit {targets[0]}",
         showlegend=True,
+        uirevision="0",  # ``uirevision`` allows zooming while live plotting
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:397]: Remove

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 14]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
         xaxis2_title="theta [rad] + virtual phase [rad]",
-        yaxis_title="Signal [a.u.]",
+        yaxis_title="MSR[V]",
     )
 
     fig2.update_layout(
-        title_text=f"Phase correction Qubit {qubits[1]}",
+        title_text=f"Phase correction Qubit {targets[1]}",
         showlegend=True,
+        uirevision="0",  # ``uirevision`` allows zooming while live plotting
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:406]: Remove

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 15]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
         xaxis1_title="theta [rad] + virtual phase[rad]",
         xaxis2_title="theta [rad] + virtual phase[rad]",
-        yaxis_title="Signal [a.u.]",
+        yaxis_title="MSR[V]",
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:409]: Restore signal

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 16]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
+    return [fig1, fig2], fitting_report
 
 
-def _update(results: CZVirtualZResults, platform: Platform, target: QubitPairId):
-    # FIXME: quick fix for qubit order
-    target = tuple(sorted(target))
-    update.virtual_phases(results.virtual_phase[target], platform, target)
+def _update(results: VirtualZResults, platform: Platform, qubit_pair: QubitPairId):
+    if qubit_pair[0] > qubit_pair[1]:
+        qubit_pair = (qubit_pair[1], qubit_pair[0])
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:417]: Previous version of more coincise

File: src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py [hunk 17]
```diff
@@ -372,50 +379,44 @@ def _plot(data: CZVirtualZData, fit: CZVirtualZResults, target: QubitPairId):
...
+    update.virtual_phases(results.virtual_phase[qubit_pair], platform, qubit_pair)
```
Comment (andrea-pasquale) [inline src/qibocal/protocols/characterization/two_qubit_interaction/native_gate_virtualz.py:418]: The `virtual_phases` method will need to be updated.

**Review by andrea-pasquale — CHANGES_REQUESTED** [review]
Thanks @Jacfomg for the feature. I'm not sure if this is ready to review, but I already reviewed it since you asked.
The idea looks good, you changed also a lot of things that were already addressed in main (going back to MSR instead of signal and several more). I know that this changes are coming from another PR but be more careful in keeping maintainability with `main`.
We should also consider to split the protocol I think.

For the future, DON'T ask for a review if the PR is not ready, just tag me.

**Comment by Jacfomg** [comment]
> Thanks @Jacfomg for the feature. I'm not sure if this is ready to review, but I already reviewed it since you asked. The idea looks good, you changed also a lot of things that were already addressed in main (going back to MSR instead of signal and several more). I know that this changes are coming from another PR but be more careful in keeping maintainability with `main`. We should also consider to split the protocol I think.
> 
> For the future, DON'T ask for a review if the PR is not ready, just tag me.

If you want to split the routine I was doing the opposite by combining CZ and iSWAP. So that answers my questions about this protocol not being what you want.

**Comment by andrea-pasquale** [comment]
> If you want to split the routine I was doing the opposite by combining CZ and iSWAP. So that answers my questions about this protocol not being what you want.

I said we should think about splitting the protocol. Why did you close the PR?
